service-providers/authenticator/src/mixnet_listener.rs

```diff
@@ -0,0 +1,275 @@
+
+    // The mixnet client that we use to send and receive packets from the mixnet
+    pub(crate) mixnet_client: nym_sdk::mixnet::MixnetClient,
+
+    // The task handle for the main loop
+    pub(crate) task_handle: TaskHandle,
+
+    // Registrations awaiting confirmation
+    pub(crate) registration_in_progres: Arc<PendingRegistrations>,
+
+    pub(crate) wireguard_gateway_data: WireguardGatewayData,
+
+    pub(crate) free_private_network_ips: Arc<PrivateIPs>,
+}
+
+impl MixnetListener {
+    pub fn new(
+        config: Config,
+        wireguard_gateway_data: WireguardGatewayData,
+        mixnet_client: nym_sdk::mixnet::MixnetClient,
+        task_handle: TaskHandle,
+    ) -> Self {
+        MixnetListener {
+            config,
+            mixnet_client,
+            task_handle,
+            registration_in_progres: Default::default(),
+            wireguard_gateway_data,
+            free_private_network_ips: Default::default(),
+        }
+    }
+
+    fn remove_from_registry(
+        &self,
+        remote_public: &PeerPublicKey,
+        gateway_client: &GatewayClient,
+    ) -> Result<()> {
+        self.wireguard_gateway_data
+            .remove_peer(gateway_client)
+            .map_err(|err| {
+                AuthenticatorError::InternalError(format!("could not remove peer: {:?}", err))
+            })?;
+        self.wireguard_gateway_data
+            .client_registry()
+            .remove(remote_public);
+        Ok(())
+    }
+
+    fn on_initial_request(
+        &mut self,
+        init_message: InitMessage,
+        reply_to: Recipient,
+    ) -> AuthenticatorHandleResult {
+        let remote_public = init_message.pub_key();
+        let nonce: u64 = fastrand::u64(..);
+        if let Some(registration_data) = self.registration_in_progres.get(&remote_public) {
+            return Ok(AuthenticatorResponse::new_pending_registration_success(
+                registration_data.value().clone(),
+                reply_to,
+            ));
+        }
+        let gateway_client_opt = if let Some(gateway_client) = self
+            .wireguard_gateway_data
+            .client_registry()
+            .get(&remote_public)
+        {
+            let mut private_ip_ref = self
+                .free_private_network_ips
+                .get_mut(&gateway_client.private_ip)
+                .ok_or(AuthenticatorError::InternalError(String::from(
+                    "could not find private IP",
+                )))?;
+            *private_ip_ref = true;
+            Some(gateway_client.clone())
+        } else {
+            None
+        };
+        if let Some(gateway_client) = gateway_client_opt {
+            self.remove_from_registry(&remote_public, &gateway_client)?;
+        }
+        let mut private_ip_ref = self
+            .free_private_network_ips
+            .iter_mut()
+            .filter(|r| **r)
+            .choose(&mut thread_rng())
+            .ok_or(AuthenticatorError::NoFreeIp)?;
```

Comment:
If this line triggers an error, does that mean we end up in a half-registered state with the IP claimed but not really finished?

Reply:
That doesn't happen here, the maximum that could happen is that they get deregistered from the previous IP they got allocated and don't get to have a new one, so they need to try later. But the problem you're asking does appear actually, with half-registered states that might never actually concluded (because of errors or bad actors). I've pushed a commit that also removes those stale registration attempts, so that they don't accumulate over time.